c++/src/kj/mutex.c++

```diff
@@ -172,6 +172,65 @@ void Mutex::lock(Exclusivity exclusivity) {
   }
 }
 
+bool Mutex::lockWithTimeout(Exclusivity exclusivity, Duration timeout) {
+  auto spec = toRelativeTimespec(timeout);
+  switch (exclusivity) {
+    case EXCLUSIVE:
+      for (;;) {
+        uint state = 0;
+        if (KJ_LIKELY(__atomic_compare_exchange_n(&futex, &state, EXCLUSIVE_HELD, false,
+                                                  __ATOMIC_ACQUIRE, __ATOMIC_RELAXED))) {
+          // Acquired.
+          break;
+        }
+
+        // The mutex is contended.  Set the exclusive-requested bit and wait.
+        if ((state & EXCLUSIVE_REQUESTED) == 0) {
+          if (!__atomic_compare_exchange_n(&futex, &state, state | EXCLUSIVE_REQUESTED, false,
+                                           __ATOMIC_RELAXED, __ATOMIC_RELAXED)) {
+            // Oops, the state changed before we could set the request bit.  Start over.
+            continue;
+          }
+
+          state |= EXCLUSIVE_REQUESTED;
+        }
+
+        auto result = syscall(SYS_futex, &futex, FUTEX_WAIT_PRIVATE, state, &spec, nullptr, 0);
+        if (result < 0) {
+          // We timed out, remove exclusive requested flag.
+          if (errno == ETIMEDOUT) {
+            __atomic_fetch_and(&futex, ~EXCLUSIVE_REQUESTED, __ATOMIC_RELEASE);
```

Comment:
Hmm, I don't think you can just remove the exclusive flag here, because there could be more than one thread waiting for an exclusive lock. I think it's safe to leave it as-is even if no one is waiting, as the only down-side is that the next thread to unlock the mutex might perform a redundant futex wakeup.

Reply:
Yeah, that's what I though. Good to know it'll work the other way.